### PR TITLE
treewide: use optionalString

### PR DESCRIPTION
**Pull request by Stunkymonkey**
###### Description of changes

currently the opportunity to use `lib.optionalString` is missed out. With this I think the code gets more readable.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/misc/distcc/default.nix [hunk 1]
```diff
@@ -27,10 +27,10 @@ let
...
           --mandir=$out/share/man
                             ${if sysconfDir == "" then "" else "--sysconfdir=${sysconfDir}"}
-                            ${if static then "LDFLAGS=-static" else ""}
-                            --with${if static == true || popt == null then "" else "out"}-included-popt
-                            --with${if avahi != null then "" else "out"}-avahi
-                            --with${if gtk3 != null then "" else "out"}-gtk
+                            ${lib.optionalString static "LDFLAGS=-static"}
+                            --with${lib.optionalString (static != true && popt != null) "out"}-included-popt
+                            --with${lib.optionalString (avahi == null) "out"}-avahi
+                            --with${lib.optionalString (gtk3 == null) "out"}-gtk
```
Comment (Ericson2314) [inline pkgs/development/tools/misc/distcc/default.nix:33]: This is better done with `lib.withFeature`

File: pkgs/tools/misc/hdf4/default.nix [hunk 2]
```diff
@@ -95,13 +96,15 @@ stdenv.mkDerivation rec {
...
-    ctest -E "${excludedTestsRegex}" --output-on-failure
-    runHook postCheck
-  '';
+  checkPhase =
+    let
+      excludedTestsRegex = lib.optionalString (excludedTests != [ ]) "(${lib.concatStringsSep "|" excludedTests})";
+    in
+    ''
+      runHook preCheck
+      ctest -E "${excludedTestsRegex}" --output-on-failure
```
Comment (Ericson2314) [inline pkgs/tools/misc/hdf4/default.nix:105]: Best fixed in a follow-up PR, but this doesn't really look right, does it?

**Review by Ericson2314 — APPROVED** [review]
I quickly skimmed it and it looked good.

File: pkgs/development/tools/misc/distcc/default.nix (hunk 1 above)
Comment (Stunkymonkey) [reply pkgs/development/tools/misc/distcc/default.nix]: thanks did not know this. is fixed

File: pkgs/tools/misc/hdf4/default.nix (hunk 2 above)
Comment (Stunkymonkey) [reply pkgs/tools/misc/hdf4/default.nix:105]: not related to this PR, therefore i will close this thread.

But I agree, that it seems odd.

**Review by Ericson2314 — APPROVED** [review]

**Comment by vcunat** [comment]
Why are there multiple regressions from this PR?

**Comment by mweinelt** [comment]
Because noone can properly review 118 changed files and the author didn't follow up nor validate their change in any meaningful way.

**Comment by Ericson2314** [comment]
Well it is very weird we had an undefined variable eval error with `thunderbird-bin`. Before pointing fingers at the author, I would like to understand why that was.

**Comment by vcunat** [comment]
Not just undefined `lib` variable (like in the other file), but also using `not` instead of `!`.

**Pull request merged**